<h1>Package scala.util.control</h1>
<h2 id="signature">
<span class="modifier_kind"> <span class="kind">package</span></span> <span class="symbol"><span class="name">control</span></span>
</h2>

<div id="template"><div id="allMembers">
<div id="types" class="types members">
<h2>Type Members</h2>
<h3 id="scala.util.control.Breaks">
<span class="modifier_kind"> <span class="kind">class </span></span><span class="symbol"><a href="breaks.html" title="A class that can be instantiated for the break control abstraction."><span class="name">Breaks</span></a><span class="result"> extends <a href="../../anyref.html" name="scala.AnyRef" id="scala.AnyRef" class="extype">AnyRef</a></span></span>
</h3>
   <div class="fullcomment"><div class="comment cmt">
<p>A class that can be instantiated for the break control abstraction. Example usage:</p>
<pre data-language="scala">val mybreaks = new Breaks
import mybreaks.{break, breakable}

breakable {
  for (...) {
    if (...) break()
  }
}</pre>
<p> Calls to break from one instantiation of <code>Breaks</code> will never target breakable objects of some other instantiation. </p>
</div></div>
<h3 id="scala.util.control.ControlThrowable">
<span class="modifier_kind"><span class="modifier">abstract </span> <span class="kind">class </span></span><span class="symbol"><a href="controlthrowable.html" title="A parent class for throwable objects intended for flow control."><span class="name">ControlThrowable</span></a><span class="result"> extends <a href="../../index.html#Throwable=Throwable" name="scala.Throwable" id="scala.Throwable" class="extmbr">Throwable</a></span></span>
</h3>
   <div class="fullcomment"><div class="comment cmt">
<p>A parent class for throwable objects intended for flow control.</p>
<p> Instances of <code>ControlThrowable</code> should not normally be caught.</p>
<p> As a convenience, <code>NonFatal</code> does not match <code>ControlThrowable</code>.</p>
<pre data-language="scala">import scala.util.control.{Breaks, NonFatal}, Breaks.{break, breakable}

breakable {
  for (v &lt;- values) {
    try {
      if (p(v)) break
      else ???
    } catch {
      case NonFatal(t) =&gt; log(t)  // can't catch a break
    }
  }
}</pre>
<p> Suppression is disabled, because flow control should not suppress an exceptional condition. Stack traces are also disabled, allowing instances of <code>ControlThrowable</code> to be safely reused.</p>
<p> Instances of <code>ControlThrowable</code> should not normally have a cause. Legacy subclasses may set a cause using <code>initCause</code>. </p>
</div></div>
<h3 id="scala.util.control.NoStackTrace">
<span class="modifier_kind"> <span class="kind">trait </span></span><span class="symbol"><a href="nostacktrace.html" title="A trait for exceptions which, for efficiency reasons, do not fill in the stack trace."><span class="name">NoStackTrace</span></a><span class="result"> extends <a href="../../index.html#Throwable=Throwable" name="scala.Throwable" id="scala.Throwable" class="extmbr">Throwable</a></span></span>
</h3>
   <div class="fullcomment">
<div class="comment cmt"><p>A trait for exceptions which, for efficiency reasons, do not fill in the stack trace. Stack trace suppression can be disabled on a global basis via a system property wrapper in <a href="../../sys/systemproperties.html" name="scala.sys.SystemProperties" id="scala.sys.SystemProperties" class="extype">scala.sys.SystemProperties</a>. </p></div>
<dl class="attributes block">
<dt>Since</dt>
<dd><p>2.8</p></dd>
<dt>Note</dt>
<dd><span class="cmt"><p>Since JDK 1.7, a similar effect can be achieved with <code>class Ex extends Throwable(..., writableStackTrace = false)</code></p></span></dd>
</dl>
</div>
</div>
<div class="values members">
<h2>Value Members</h2>
<h3 id="scala.util.control.Breaks">
<span class="modifier_kind"> <span class="kind">object </span></span><span class="symbol"><a href="breaks%24.html" title="An object that can be used for the break control abstraction."><span class="name">Breaks</span></a><span class="result"> extends <a href="breaks.html" name="scala.util.control.Breaks" id="scala.util.control.Breaks" class="extype">Breaks</a></span></span>
</h3>
   <div class="fullcomment"><div class="comment cmt">
<p>An object that can be used for the break control abstraction. Example usage:</p>
<pre data-language="scala">import Breaks.{break, breakable}

breakable {
  for (...) {
    if (...) break
  }
}</pre>
</div></div>
<h3 id="scala.util.control.Exception">
<span class="modifier_kind"> <span class="kind">object </span></span><span class="symbol"><a href="exception%24.html" title="Classes representing the components of exception handling."><span class="name">Exception</span></a></span>
</h3>
   <div class="fullcomment"><div class="comment cmt">
<p>Classes representing the components of exception handling.</p>
<p> Each class is independently composable.</p>
<p> This class differs from <a href="../try.html" name="scala.util.Try" id="scala.util.Try" class="extype">scala.util.Try</a> in that it focuses on composing exception handlers rather than composing behavior. All behavior should be composed first and fed to a <a href="exception%24%24catch.html" name="scala.util.control.Exception.Catch" id="scala.util.control.Exception.Catch" class="extype">Catch</a> object using one of the <code>opt</code>, <code>either</code> or <code>withTry</code> methods. Taken together the classes provide a DSL for composing catch and finally behaviors.</p>
<h5> Examples </h5>
<p> Create a <code>Catch</code> which handles specified exceptions.</p>
<pre data-language="scala">import scala.util.control.Exception._
import java.net._

val s = "http://www.scala-lang.org/"

// Some(http://www.scala-lang.org/)
val x1: Option[URL] = catching(classOf[MalformedURLException]) opt new URL(s)

// Right(http://www.scala-lang.org/)
val x2: Either[Throwable,URL] =
  catching(classOf[MalformedURLException], classOf[NullPointerException]) either new URL(s)

// Success(http://www.scala-lang.org/)
val x3: Try[URL] = catching(classOf[MalformedURLException], classOf[NullPointerException]) withTry new URL(s)

val defaultUrl = new URL("http://example.com")
//  URL(http://example.com) because htt/xx throws MalformedURLException
val x4: URL = failAsValue(classOf[MalformedURLException])(defaultUrl)(new URL("htt/xx"))</pre>
<p> Create a <code>Catch</code> which logs exceptions using <code>handling</code> and <code>by</code>.</p>
<pre data-language="scala">def log(t: Throwable): Unit = t.printStackTrace

val withThrowableLogging: Catch[Unit] = handling(classOf[MalformedURLException]) by (log)

def printUrl(url: String) : Unit = {
  val con = new URL(url) openConnection()
  val source = scala.io.Source.fromInputStream(con.getInputStream())
  source.getLines.foreach(println)
}

val badUrl = "htt/xx"
// Prints stacktrace,
//   java.net.MalformedURLException: no protocol: htt/xx
//     at java.net.URL.&lt;init&gt;(URL.java:586)
withThrowableLogging { printUrl(badUrl) }

val goodUrl = "http://www.scala-lang.org/"
// Prints page content,
//   &lt;!DOCTYPE html&gt;
//   &lt;html&gt;
withThrowableLogging { printUrl(goodUrl) }</pre>
<p> Use <code>unwrapping</code> to create a <code>Catch</code> that unwraps exceptions before rethrowing.</p>
<pre data-language="scala">class AppException(cause: Throwable) extends RuntimeException(cause)

val unwrappingCatch: Catch[Nothing] = unwrapping(classOf[AppException])

def calcResult: Int = throw new AppException(new NullPointerException)

// Throws NPE not AppException,
//   java.lang.NullPointerException
//     at .calcResult(&lt;console&gt;:17)
val result = unwrappingCatch(calcResult)</pre>
<p> Use <code>failAsValue</code> to provide a default when a specified exception is caught.</p>
<pre data-language="scala">val inputDefaulting: Catch[Int] = failAsValue(classOf[NumberFormatException])(0)
val candidatePick = "seven" // scala.io.StdIn.readLine()

// Int = 0
val pick = inputDefaulting(candidatePick.toInt)</pre>
<p> Compose multiple <code>Catch</code>s with <code>or</code> to build a <code>Catch</code> that provides default values varied by exception.</p>
<pre data-language="scala">val formatDefaulting: Catch[Int] = failAsValue(classOf[NumberFormatException])(0)
val nullDefaulting: Catch[Int] = failAsValue(classOf[NullPointerException])(-1)
val otherDefaulting: Catch[Int] = nonFatalCatch withApply(_ =&gt; -100)

val combinedDefaulting: Catch[Int] = formatDefaulting or nullDefaulting or otherDefaulting

def p(s: String): Int = s.length * s.toInt

// Int = 0
combinedDefaulting(p("tenty-nine"))

// Int = -1
combinedDefaulting(p(null: String))

// Int = -100
combinedDefaulting(throw new IllegalStateException)

// Int = 22
combinedDefaulting(p("11"))</pre>
</div></div>
<h3 id="scala.util.control.NoStackTrace">
<span class="modifier_kind"> <span class="kind">object </span></span><span class="symbol"><a href="nostacktrace%24.html" title=""><span class="name">NoStackTrace</span></a><span class="result"> extends <span name="java.io.Serializable" class="extype">java.io.Serializable</span></span></span>
</h3>
   <h3 id="scala.util.control.NonFatal">
<span class="modifier_kind"> <span class="kind">object </span></span><span class="symbol"><a href="nonfatal%24.html" title="Extractor of non-fatal Throwables."><span class="name">NonFatal</span></a></span>
</h3>
   <div class="fullcomment"><div class="comment cmt">
<p>Extractor of non-fatal Throwables. Will not match fatal errors like <code>VirtualMachineError</code> (for example, <code>OutOfMemoryError</code> and <code>StackOverflowError</code>, subclasses of <code>VirtualMachineError</code>), <code>ThreadDeath</code>, <code>LinkageError</code>, <code>InterruptedException</code>, <code>ControlThrowable</code>.</p>
<p>Note that <a href="controlthrowable.html" name="scala.util.control.ControlThrowable" id="scala.util.control.ControlThrowable" class="extype">scala.util.control.ControlThrowable</a>, an internal Throwable, is not matched by <code>NonFatal</code> (and would therefore be thrown).</p>
<p>For example, all harmless Throwables can be caught by:</p>
<pre data-language="scala">try {
  // dangerous stuff
} catch {
  case NonFatal(e) =&gt; log.error(e, "Something not that bad.")
 // or
  case e if NonFatal(e) =&gt; log.error(e, "Something not that bad.")
}</pre>
</div></div>
<h3 id="scala.util.control.TailCalls">
<span class="modifier_kind"> <span class="kind">object </span></span><span class="symbol"><a href="tailcalls%24.html" title="Methods exported by this object implement tail calls via trampolining."><span class="name">TailCalls</span></a></span>
</h3>
   <div class="fullcomment"><div class="comment cmt">
<p>Methods exported by this object implement tail calls via trampolining. Tail calling methods have to return their result using <code>done</code> or call the next method using <code>tailcall</code>. Both return a <code>TailRec</code> object. The result of evaluating a tailcalling function can be retrieved from a <code>Tailrec</code> value using method <code>result</code>. Implemented as described in "Stackless Scala with Free Monads" <a href="http://blog.higher-order.com/assets/trampolines.pdf" target="_blank">http://blog.higher-order.com/assets/trampolines.pdf</a></p>
<p> Here's a usage example:</p>
<pre data-language="scala">import scala.util.control.TailCalls._

def isEven(xs: List[Int]): TailRec[Boolean] =
  if (xs.isEmpty) done(true) else tailcall(isOdd(xs.tail))

def isOdd(xs: List[Int]): TailRec[Boolean] =
 if (xs.isEmpty) done(false) else tailcall(isEven(xs.tail))

isEven((1 to 100000).toList).result

def fib(n: Int): TailRec[Int] =
  if (n &lt; 2) done(n) else for {
    x &lt;- tailcall(fib(n - 1))
    y &lt;- tailcall(fib(n - 2))
  } yield (x + y)

fib(40).result</pre>
</div></div>
</div>
</div></div>
<div class="_attribution">
  <p class="_attribution-p">
    &copy; 2002-2019 EPFL, with contributions from Lightbend.<br>Licensed under the Apache License, Version 2.0.<br>
    <a href="https://www.scala-lang.org/api/2.13.0/scala/util/control/index.html" class="_attribution-link">https://www.scala-lang.org/api/2.13.0/scala/util/control/index.html</a>
  </p>
</div>
